Modules/grpmodule.c

```diff
@@ -119,16 +120,56 @@ grp_getgrgid_impl(PyObject *module, PyObject *id)
         }
         Py_DECREF(py_int_id);
     }
+#ifdef HAVE_GETGRGID_R
+    Py_BEGIN_ALLOW_THREADS
+    int status;
+    long bufsize = sysconf(_SC_GETGR_R_SIZE_MAX);
+    struct group grp;
+
+    buf = PyMem_RawMalloc(bufsize);
+
+    while (1) {
+        status = getgrgid_r(gid, &grp, buf, bufsize, &p);
+        if (p != NULL || status != ERANGE) {
+            break;
+        }
+        if (bufsize > PY_SSIZE_T_MAX) {
+            PyErr_NoMemory();
```

Comment:
The MemoryError will be overridden by a KeyError below. You need to avoid this. For example use a special flag or check `PyErr_Occurred()` before setting a KeyError.